Frameworks/Starboard/EbrStorageFile.cpp

```diff
@@ -0,0 +1,351 @@
+    RETURN_NULL_IF_FAILED(
+        GetActivationFactory(Wrappers::HStringReference(RuntimeClass_Windows_Storage_AccessCache_StorageApplicationPermissions).Get(),
+                             &storageApplicationPermissionsStatics));
+
+    ComPtr<IStorageItemAccessList> accessList;
+    RETURN_NULL_IF_FAILED(storageApplicationPermissionsStatics->get_FutureAccessList(&accessList));
+
+    // Try to grab the item based on id.
+    ComPtr<IStorageFile> storageFile;
+    RETURN_NULL_IF_FAILED(AwaitResult(accessList, &IStorageItemAccessList::GetFileAsync, idHString.Get(), &storageFile));
+
+    bool shouldAllowWrite = ((mode & _O_WRONLY) != 0) ||
+                            ((mode & _O_RDWR) != 0);
+    
+    FileAccessMode openMode = shouldAllowWrite ? FileAccessMode_ReadWrite : FileAccessMode_Read;
+
+    ComPtr<IRandomAccessStream> randomAccessStream;
+    RETURN_NULL_IF_FAILED(AwaitResult(storageFile, &IStorageFile::OpenAsync, openMode, &randomAccessStream));
+
+    bool shouldAppend = ((mode & _O_APPEND) != 0);
+
+    return std::make_shared<EbrStorageFile>(
+        std::move(storageFile),
+        std::move(randomAccessStream),
+        shouldAppend,
+        shouldAllowWrite);
+}
+
+EbrStorageFile::EbrStorageFile(
+        ComPtr<IStorageFile>&& storageFile, 
+        ComPtr<IRandomAccessStream>&& randomAccessStream,
+        bool shouldAppend,
+        bool shouldAllowWrite) : 
+            m_storageFile(storageFile), 
+            m_randomAccessStream(randomAccessStream), 
+            m_shouldAppend(shouldAppend), 
+            m_shouldAllowWrite(shouldAllowWrite) {
+
+    if (shouldAppend) {
+        Lseek(0, SEEK_END);
+    }
+}
+
+int EbrStorageFile::Tell() {
+    return Lseek(0, SEEK_CUR);
+}
+
+int EbrStorageFile::Flush() {
+    std::lock_guard<std::recursive_mutex> lock(m_lock);
+
+    unsigned long long currentPosition{};
+    if (FAILED(m_randomAccessStream->get_Position(&currentPosition))) {
+        return -1;
+    }
+
+    ComPtr<IOutputStream> outStream;
+    if (FAILED(m_randomAccessStream->GetOutputStreamAt(currentPosition, &outStream))) {
+        return -1;
+    }
+
+    boolean success = false;
+    if (FAILED(AwaitResult(outStream, &IOutputStream::FlushAsync, &success))) {
+        return -1;
+    }
+
+    return success ? 0 : -1;
+}
+
+int EbrStorageFile::Stat(struct stat* ret) {
+    std::lock_guard<std::recursive_mutex> lock(m_lock);
+
+    if (!ret) {
+        return -1;
+    }
+
+    struct stat toReturn{};
+
+    ComPtr<IStorageItem> storageItem;
+    if (FAILED(m_storageFile.As(&storageItem))) {
+        return -1;
+    }
+
+    ComPtr<IBasicProperties> basicProperties;
+    if (FAILED(AwaitResult(storageItem, &IStorageItem::GetBasicPropertiesAsync, &basicProperties))) {
+        return -1;
+    }
+
+    unsigned long long size{};
+    if (FAILED(basicProperties->get_Size(&size))) {
+        return -1;
+    }
+
+    // TODO 1099: Add more properites that we can fill in stat.
```

Comment:
Yes, please remove TODO.

---
In reply to: [84801640](https://github.com/Microsoft/WinObjC/pull/1215#discussion_r84801640) [](ancestors = 84801640)